crates/ide/src/completion/complete_postfix/format_like.rs

```diff
@@ -0,0 +1,325 @@
+//! Postfix completion for `format`-like strings.
+//!
+//! `"Result {result} is {2 + 2}"` is expanded to the `"Result {} is {}", result, 2 + 2`.
+//!
+//! The following postfix snippets are available:
+//!
+//! - `format` -> `format!(...)`
+//! - `panic` -> `panic!(...)`
+//! - `println` -> `println!(...)`
+//! - `log`:
+//!   + `logd` -> `log::debug!(...)`
+//!   + `logt` -> `log::trace!(...)`
+//!   + `logi` -> `log::info!(...)`
+//!   + `logw` -> `log::warn!(...)`
+//!   + `loge` -> `log::error!(...)`
+
+use super::postfix_snippet;
+use crate::completion::{
+    completion_config::SnippetCap, completion_context::CompletionContext,
+    completion_item::Completions,
+};
+use syntax::ast;
+
+pub(super) fn add_format_like_completions(
+    acc: &mut Completions,
+    ctx: &CompletionContext,
+    dot_receiver: &ast::Expr,
+    cap: SnippetCap,
+    receiver_text: &str,
+) {
+    if !is_string_literal(receiver_text) {
+        // It's not a string literal, do not parse input.
+        return;
+    }
+
+    let input = &receiver_text[1..receiver_text.len() - 1];
+
+    let mut parser = FormatStrParser::new(input);
+
+    if parser.parse().is_ok() {
+        for kind in PostfixKind::all_suggestions() {
+            let snippet = parser.into_suggestion(*kind);
+            let (label, detail) = kind.into_description();
+
+            postfix_snippet(ctx, cap, &dot_receiver, label, detail, &snippet).add_to(acc);
+        }
+    }
+}
+
+/// Checks whether provided item is a string literal.
+fn is_string_literal(item: &str) -> bool {
+    if item.len() < 2 {
+        return false;
+    }
+    item.starts_with("\"") && item.ends_with("\"")
+}
+
+/// Parser for a format-like string. It is more allowing in terms of string contents,
+/// as we expect variable placeholders to be filled with expressions.
+#[derive(Debug)]
+pub struct FormatStrParser {
+    input: String,
+    output: String,
+    extracted_expressions: Vec<String>,
+    state: State,
+    parsed: bool,
+}
+
+#[derive(Debug, Clone, Copy)]
+pub enum PostfixKind {
+    Format,
+    Panic,
+    Println,
+    LogDebug,
+    LogTrace,
+    LogInfo,
+    LogWarn,
+    LogError,
+}
```

Comment:
When all enum variants are the same, it's often better to use a data table instead:

```rust
static KINDS &[(&str, &str)] = &[
  ("fmt", "format!"),
  ("panic", "panic!"),
];
```

I think the current code is OK, but I would guess that using an array and some helper functions wold shorten it.